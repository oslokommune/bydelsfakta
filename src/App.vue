<template>
  <div id="app" class="app" tabindex="-1" :class="{ menuIsOpen, navigationIsOpen }">
    <the-navigation-drawer />
    <div class="app__content">
      <the-navigation-topbar />
      <main>
        <router-view />
      </main>
      <the-footer />
    </div>
    <modal v-if="isOpen" @close="isOpen = false" />
  </div>
</template>

<script>
import { mapState, mapActions } from 'vuex';
import TheNavigationDrawer from './components/TheNavigationDrawer';
import TheNavigationTopbar from './components/TheNavigationTopbar';
import TheFooter from './components/TheFooter';
import Modal from './components/Modal';

export default {
  name: 'App',

  components: { TheNavigationTopbar, TheNavigationDrawer, TheFooter, Modal },

  data: () => ({
    isOpen: false,
  }),

  metaInfo() {
    return {
      title: 'Bydelsfakta – Visualisering av statistikk om befolkning, levekår og boforhold',
    };
  },

  computed: {
    ...mapState(['menuIsOpen', 'navigationIsOpen', 'ie11']),
  },

  watch: {
    $route(to) {
      if (to.params.district !== undefined) {
        this.addDistrict({ district: to.params.district, pushRoute: false });
        document.getElementById('app').focus();
      }
    },
  },

  mounted() {
    if ('ontouchstart' in document.documentElement) {
      this.setTouchDevice(true);
    }
    if (!!window.MSInputMethodContext && !!document.documentMode) {
      this.isOpen = true;
      this.setIE11Compatibility(true);
    }
  },

  created() {
    if (process.env.NODE_ENV === 'production') {
      const envs = JSON.parse(window.__GLOBAL_ENVS__);
      this.setProductionMode(envs.VUE_APP_PRODUCTION_DATA);
      document.getElementById('bydelsfakta-globals').remove();
    } else {
      this.setProductionMode(process.env.VUE_APP_PRODUCTION_DATA);
    }

    if (this.$route.name === 'Home') {
      this.addDistrict({ district: 'alle', pushRoute: false });
    }
    if (this.$route.params.district === undefined) {
      return;
    }
    this.addDistrict({ district: this.$route.params.district, pushRoute: true });
  },

  methods: {
    ...mapActions(['addDistrict', 'setTouchDevice', 'setIE11Compatibility', 'setProductionMode']),
  },
};

// Using a class on body to determine how to style focus states
document.body.addEventListener('mousedown', () => {
  document.body.classList.add('using-mouse');
});
document.body.addEventListener('keydown', () => {
  document.body.classList.remove('using-mouse');
});
</script>

<style lang="scss">
.app:focus {
  outline: none;
}
</style>
