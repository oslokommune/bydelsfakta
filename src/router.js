import Vue from 'vue';
import Router from 'vue-router';

import { topicNames, disabledTopics } from './config/topics';
import allDistricts from './config/allDistricts';

const District = () => import('./views/District');
const Topic = () => import('./views/Topic');
const NotFound = () => import('./views/NotFound');

Vue.use(Router);

export const routes = [
  {
    path: '/',
    name: 'Home',
    component: District,
  },
  {
    path: '/bydel/:district?',
    name: 'District',
    component: District,
    props: true,
  },
  {
    path: '/bydel/:district/:topic',
    name: 'Topic',
    component: Topic,
    props: true,
  },
  {
    path: '*',
    name: 'NotFound',
    component: NotFound,
  },
];

const router = new Router({
  mode: 'history',
  routes,
});

// Temp while we find out what our homepage should look like
router.beforeEach((to, from, next) => {
  if (to.params.topic) {
    if (!topicNames.find(name => name === to.params.topic) || disabledTopics.includes(to.params.topic)) {
      next({ name: 'NotFound', params: [to.path] });
    }
  }

  if (to.params.district) {
    const districts = to.params.district.split('-');
    if (districts[0] === 'alle') {
      next();
    } else if (districts.length === 1) {
      if (allDistricts.find(district => district.uri === districts[0])) {
        next();
      } else if (allDistricts.find(district => district.key === districts[0])) {
        next();
      } else {
        next({ name: 'NotFound', params: [to.path] });
      }
    } else if (districts.length > 1) {
      const errors = [];
      districts.forEach(district => {
        if (!allDistricts.find(item => item.key === district)) {
          errors.push(district);
        }
      });
      if (errors.length > 0) {
        next({ name: 'NotFound', params: [to.path] });
      }
    }
  }

  if (to.params.topic !== from.params.topic || (to.name === 'Topic' && from.name === 'District')) {
    window.scroll(0, 0);
  }

  next();
});

export default router;
