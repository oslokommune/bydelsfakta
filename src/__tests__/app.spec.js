import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';
import App from '../App';
import router from '../router';
import clickOutside from '../directives/clickOutside';
import store from '../store';
import setupI18n from '../i18n';

global.scroll = jest.fn();
window.scroll = jest.fn();

const i18n = setupI18n();

describe('App', () => {
  let wrapper = null;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(router);
    localVue.use(Vuex);
    localVue.directive('click-outside', clickOutside);
    wrapper = shallowMount(App, {
      localVue,
      router,
      store,
      i18n,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('renders app-component and finds id #app', () => {
    expect(wrapper.classes('app')).toBe(true);
  });

  test('shallowmounts app', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
