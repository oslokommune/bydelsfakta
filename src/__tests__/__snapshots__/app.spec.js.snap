// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App shallowmounts app 1`] = `
<div
  class="app"
>
  <the-navigation-drawer-stub />
   
  <div
    class="app__content"
  >
    <the-navigation-topbar-stub />
     
    <main>
      <router-view-stub
        name="default"
      />
    </main>
     
    <the-footer-stub />
  </div>
</div>
`;
